qiskit/result/result.py

```diff
@@ -170,23 +170,47 @@ def get_counts(self, experiment=None):
                 according to the registers in circuit (e.g. ``0100 1110``).
                 The string is little-endian (cr[0] on the right hand side).
 
+            List[dict[str:int]]: a list of dictionaries for each experiment
+                executed. Every dictionary has the counts for each qubit, with
+                the keys containing a string in binary format and separated
+                according to the registers in circuit (e.g. ``0100 1110``).
+                The string is little-endian (cr[0] on the right hand side).
+
         Raises:
             QiskitError: if there are no counts for the experiment.
         """
-        exp = self._get_experiment(experiment)
-        try:
-            header = exp.header.to_dict()
-        except (AttributeError, QiskitError):  # header is not available
-            header = None
-
-        if 'counts' in self.data(experiment).keys():
-            return postprocess.format_counts(self.data(experiment)['counts'],
-                                             header)
-        elif 'statevector' in self.data(experiment).keys():
-            vec = postprocess.format_statevector(self.data(experiment)['statevector'])
-            return state_to_counts(vec)
+        if experiment is None and len(self.results) > 1:
+            dict_list = []
+            for i in range(len(self.results)):
+                exp = self._get_experiment(i)
+                try:
+                    header = exp.header.to_dict()
+                except (AttributeError, QiskitError):  # header is not available
+                    header = None
+
+                if 'counts' in self.data(i).keys():
+                    dict_list.append(postprocess.format_counts(self.data(i)['counts'], header))
+                elif 'statevector' in self.data(i).keys():
+                    vec = postprocess.format_statevector(self.data(i)['statevector'])
+                    dict_list.append(state_to_counts(vec))
+                else:
+                    raise QiskitError(('No counts for experiment '+'"{'+str(i)+'}"').format(i))
+            return dict_list
         else:
-            raise QiskitError('No counts for experiment "{0}"'.format(experiment))
+            exp = self._get_experiment(experiment)
+            try:
+                header = exp.header.to_dict()
+            except (AttributeError, QiskitError):  # header is not available
+                header = None
+
+            if 'counts' in self.data(experiment).keys():
+                return postprocess.format_counts(self.data(experiment)['counts'],
+                                                 header)
+            elif 'statevector' in self.data(experiment).keys():
+                vec = postprocess.format_statevector(self.data(experiment)['statevector'])
+                return state_to_counts(vec)
+            else:
+                raise QiskitError('No counts for experiment "{0}"'.format(experiment))
```

Comment:
Instead of repeating the same code, you can actually just convert single `experiment` input into an iterable, something like:
```
exp_keys = range(len(self.results)) if experiment is None and len(self.results) > 1 else {experiment}
for key in exp_keys:
    <same code as lines 185-197>
```
And return the first in the `dict_list` if it has only 1 item.

Reply:
Ok I can try to do it in this way. I left the same code for the else in order to deal also with the case when the method is called by a job that run multiple circuits but with a particular experiment, something like:

`counts = job.result().get_counts(1)`

Anyway I think that I can handle also this case with the code you told me.

Going to try later this afternoon, if you could ensure me that also this case is covered by the solution you gave me, it would be great!

Reply:
https://github.com/Qiskit/qiskit-terra/blob/4990730a74ab7ae5fa72e494f143cd815ab1e8ae/qiskit/result/result.py#L163-L165

I tried to convert the single experiment to an iterable but it is not possible due to the fact that experiment can also be a QuantumCircuit or Schedule more than just an int or None. 

I think that the solution of duplicating the code is the only one, it is also convenient to handle the last exception when we need to print the index of the experiment with no counts:

`raise QiskitError(('No counts for experiment '+'"{'+str(i)+'}"').format(i))`

Reply:
If you look at the old code, the `raise` was 
```
 raise QiskitError('No counts for experiment "{0}"'.format(experiment))
```
This formats the `experiment` argument passed in, which may be an `int`, or  `QuantumCircuit`, or `Schedule`. So if you use the same code but change `experiment` to `key`:
```
 raise QiskitError('No counts for experiment "{0}"'.format(key))
```
it'll also format the string nicely regardless the type of `key`.

Reply:
That is exactly what I did, if you look at my last commit I have pushed the updated version with your suggestions

Reply:
I'm still seeing 
```
raise QiskitError(('No counts for experiment '+'"{'+str(key)+'}"').format(key))
```
which has the extra `{` and `}` in the formatted string compared to the old format (where `{0}` is replaced by `experiment`).

Reply:
You are right, I am sorry, thought I updated it last night but evidently I didn't, going to fix this too while adding the test and the CHANGELOG.